File: lua/problem_56.lua
local M = {}

function M.graph_colorable(adj, k)
    local n = #adj
    local colors = {}
    for i = 1, n do colors[i] = 0 end
    local function bt(node)
        if node > n then return true end
        for c = 1, k do
            local ok = true
            for nb = 1, n do
                if adj[node][nb] == 1 and colors[nb] == c then ok = false; break end
            end
            if ok then
                colors[node] = c
                if bt(node + 1) then return true end
                colors[node] = 0
            end
        end
        return false
    end
    return bt(1)
end

function M.problem56()
    print("Problem 56:")
    local t = {{0,1,1},{1,0,1},{1,1,0}}
    assert(    M.graph_colorable(t, 3))
    assert(not M.graph_colorable(t, 2))
    print("All Tests Passed")
end

return M

File: rust/problem_56.rs
pub fn graph_colorable(adj: &Vec<Vec<i32>>, k: usize) -> bool {
    let n = adj.len();
    let mut colors = vec![0usize; n];
    fn bt(node: usize, n: usize, adj: &Vec<Vec<i32>>, k: usize, colors: &mut Vec<usize>) -> bool {
        if node == n { return true; }
        for c in 1..=k {
            if (0..n).all(|nb| adj[node][nb] == 0 || colors[nb] != c) {
                colors[node] = c;
                if bt(node+1, n, adj, k, colors) { return true; }
                colors[node] = 0;
            }
        }
        false
    }
    bt(0, n, adj, k, &mut colors)
}

pub fn problem56() {
    println!("Problem 56:");
    let t = vec![vec![0,1,1],vec![1,0,1],vec![1,1,0]];
    assert!( graph_colorable(&t, 3));
    assert!(!graph_colorable(&t, 2));
    println!("All Tests Passed");
}
